src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -237,12 +251,18 @@ RouterExecutorRun(QueryDesc *queryDesc, ScanDirection direction, long count, Tas
  * also generate warnings for individual placement failures.
  */
 static int32
-ExecuteDistributedModify(Task *task)
+ExecuteDistributedModify(Query *query, Task *task)
 {
 	int32 affectedTupleCount = -1;
 	ListCell *taskPlacementCell = NULL;
 	List *failedPlacementList = NIL;
 	ListCell *failedPlacementCell = NULL;
+	StringInfo queryString = makeStringInfo();
+
+	ExecuteFunctions(query);
+	DeparseShardQuery(query, task, queryString);
+	elog(DEBUG4, "old query: %s", task->queryString);
+	elog(DEBUG4, "new query: %s", queryString->data);
```

Comment:
I think we should have a safety check here, verifying that no volatile/stable functions are present anymore.